### PR TITLE
Se/add support proxy httpapi compact

### PR DESCRIPTION
**Pull request by ewertons**
This is a PR that replaces https://github.com/Azure/azure-c-shared-utility/pull/446
Adding fixes and unit tests.

### CODE REVIEW

File: adapters/httpapi_compact.c [hunk 1]
```diff
@@ -1341,6 +1356,75 @@ HTTPAPI_RESULT HTTPAPI_SetOption(HTTP_HANDLE handle, const char* optionName, con
...
+                (proxy_options->username != proxy_options->password))
+        {
+            LogError("Only one of username and password for proxy settings was NULL");
+            result = HTTPAPI_ERROR;
+        }
+        else
+        {
+
+            /* Workaround: xio interface is already created when HTTPAPI_CreateConnection is call without proxy support
+             * need to destroy the interface and create a new one with proxy information
```
Comment (danewalton) [inline adapters/httpapi_compact.c:1380]: pleading ignorance here: so nothing gets lost in the destruction? Maybe some option a user would set that doesn't get carried over?

File: adapters/httpapi_compact.c [hunk 2]
```diff
@@ -1341,6 +1356,75 @@ HTTPAPI_RESULT HTTPAPI_SetOption(HTTP_HANDLE handle, const char* optionName, con
...
+                tlsio_config.hostname = http_instance->hostName;
+                tlsio_config.port = 443;
+                tlsio_config.underlying_io_interface =  http_proxy_io_get_interface_description();
+                tlsio_config.underlying_io_parameters = &proxy_config;
+
+                http_instance->xio_handle = xio_create(platform_get_default_tlsio(), (void*)&tlsio_config);
+
+                if (http_instance->xio_handle == NULL)
+                {
+                    LogError("Fail to create xio handle with proxy configuration");
```
Comment (danewalton) [inline adapters/httpapi_compact.c:1408]: nit:
```suggestion
                    LogError("Failed to create xio handle with proxy configuration");
```

File: adapters/httpapi_compact.c (hunk 1 above)
Comment (danewalton) [reply adapters/httpapi_compact.c:1380]: maybe line 1413 is doing that?

Comment (ewertons) [reply adapters/httpapi_compact.c:1380]: yes, retrieve options and feedoptions are guaranteeing previous settings are preserved.

File: adapters/httpapi_compact.c [hunk 3]
```diff
@@ -1425,11 +1509,31 @@ HTTPAPI_RESULT HTTPAPI_CloneOption(const char* optionName, const void* value, co
...
+    }
     else
     {
         /*Codes_SRS_HTTPAPI_COMPACT_21_071: [ If the HTTP do not support the optionName, the HTTPAPI_CloneOption shall return HTTPAPI_INVALID_ARG. ]*/
         result = HTTPAPI_INVALID_ARG;
         LogInfo("unknown option %s", optionName);
     }
     return result;
-}
+}
```
[the remark this answers is not in the document]
Comment (ewertons) [reply adapters/httpapi_compact.c:1539]: Fixed.

File: adapters/httpapi_compact.c (hunk 2 above)
Comment (ewertons) [reply adapters/httpapi_compact.c:1408]: Fixed. Thanks.

File: adapters/httpapi_compact.c [hunk 4]
```diff
@@ -1341,12 +1356,82 @@ HTTPAPI_RESULT HTTPAPI_SetOption(HTTP_HANDLE handle, const char* optionName, con
...
+
+        if (proxy_options->host_address == NULL)
+        {
+            LogError("NULL host_address in proxy options");
+            result = HTTPAPI_ERROR;
+        }
+        else if (((proxy_options->username == NULL) || (proxy_options->password == NULL)) &&
+                (proxy_options->username != proxy_options->password))
+        {
+            LogError("Only one of username and password for proxy settings was NULL");
```
Comment (ericwol-msft) [inline adapters/httpapi_compact.c:1373]:
```suggestion
            LogError("Either username or password for proxy settings was NULL");
```

File: tests/httpapicompact_ut/httpapicompact_ut.c [hunk 5]
```diff
@@ -60,6 +60,14 @@ void my_gballoc_free(void* ptr)
     free(ptr);
 }
 
+int my_mallocAndStrcpy_s(char** destination, const char* source)
```
Comment (ericwol-msft) [inline tests/httpapicompact_ut/httpapicompact_ut.c:63]:
```suggestion
int custom_mallocAndStrcpy_s(char** destination, const char* source)
```

Comment (ewertons) [reply tests/httpapicompact_ut/httpapicompact_ut.c:63]: This is a function that has been named like that for quite a while, and used everywhere in the track 1 SDK. Changing the name would mean a huge refactoring  of the code. I think we should postpone that change.

File: adapters/httpapi_compact.c (hunk 4 above)
Comment (ewertons) [reply adapters/httpapi_compact.c:1373]: Oops, I didn't see your comment until after I merged. I can submit another one to change the text.

Comment (ericwol-msft) [reply adapters/httpapi_compact.c:1373]: minor nit, leave it as is

**Pull request merged**